# GitHub e Base

In [1]:
import os

# 1. Define o repositório e o caminho local no Colab
repositorio = 'https://github.com/fusieger/ml-cybercrime-public-safety.git'
caminho_repositorio = '/content/ml-cybercrime-public-safety'

# 2. Clona o repositório caso ele ainda não exista no ambiente
if not os.path.exists(caminho_repositorio):
    !git clone $repositorio

# 3. Acessa a pasta principal do projeto
os.chdir(caminho_repositorio)

print(f'Repositório carregado com sucesso em: {os.getcwd()}')

Cloning into 'ml-cybercrime-public-safety'...
remote: Enumerating objects: 296, done.
remote: Counting objects: 100% (296/296), done.
remote: Compressing objects: 100% (253/253), done.
remote: Total 296 (delta 110), reused 0 (delta 0), pack-reused 0 (from 0)
Receiving objects: 100% (296/296), 29.17 MiB | 5.64 MiB/s, done.
Resolving deltas: 100% (110/110), done.
Repositório carregado com sucesso em: /content/ml-cybercrime-public-safety


In [2]:
import pandas as pd
import os

# 1. Define as pastas principais dos dados processados no repositório
pasta_fraude = os.path.join(
    '/content/ml-cybercrime-public-safety',
    'data',
    'processed',
    'sesp-es',
    'estelionato-fraude'
)

pasta_populacao = os.path.join(
    '/content/ml-cybercrime-public-safety',
    'data',
    'processed',
    'ibge',
    'populacao'
)

# 2. Mapeia os caminhos exatos dos arquivos
caminho_fraude_2025 = os.path.join(
    pasta_fraude,
    'estelionato-e-fraude-2025-es-clean.csv'
)

caminho_fraude_geral = os.path.join(
    pasta_fraude,
    'estelionato-e-fraude-es-clean.csv'
)

caminho_populacao = os.path.join(
    pasta_populacao,
    'populacao-es-clean.csv'
)

# 3. Valida se os arquivos necessários estão disponíveis
arquivos_necessarios = [
    caminho_fraude_2025,
    caminho_fraude_geral,
    caminho_populacao
]

for arquivo in arquivos_necessarios:
    if not os.path.exists(arquivo):
        raise FileNotFoundError(f'Arquivo não encontrado: {arquivo}')

# 4. Carrega os arquivos com a codificação e separadores corretos
df_fraude_2025 = pd.read_csv(
    caminho_fraude_2025,
    encoding='latin1',
    sep=';'
)

df_fraude_geral = pd.read_csv(
    caminho_fraude_geral,
    encoding='latin1',
    sep=';'
)

df_populacao = pd.read_csv(
    caminho_populacao,
    encoding='latin1',
    sep=';',
    thousands='.'
)

print("Bases de segurança pública e demografia carregadas com sucesso!")


Bases de segurança pública e demografia carregadas com sucesso!


In [3]:
from IPython.display import display

print("=== DIMENSÕES DAS BASES ===")
print(f"Fraude 2025:  {df_fraude_2025.shape[0]:,} registros | {df_fraude_2025.shape[1]} colunas")
print(f"Fraude Geral: {df_fraude_geral.shape[0]:,} registros | {df_fraude_geral.shape[1]} colunas")
print(f"População:    {df_populacao.shape[0]:,} registros | {df_populacao.shape[1]} colunas")

print("\n=== PERÍODO DOS DADOS ===")
print(
    f"Fraude Geral: "
    f"{df_fraude_geral['ano'].min()} a {df_fraude_geral['ano'].max()}"
)
print(
    f"Fraude 2025: "
    f"{df_fraude_2025['ano'].min()} a {df_fraude_2025['ano'].max()}"
)
print(
    f"População: "
    f"{df_populacao['ano'].min()} a {df_populacao['ano'].max()}"
)

print("\n=== MUNICÍPIOS ===")
print(f"Fraude 2025:  {df_fraude_2025['municipio'].nunique()}")
print(f"Fraude Geral: {df_fraude_geral['municipio'].nunique()}")
print(f"População:    {df_populacao['municipio'].nunique()}")

print("\n=== VALORES AUSENTES ===")
print(f"Fraude 2025:  {df_fraude_2025.isna().sum().sum()}")
print(f"Fraude Geral: {df_fraude_geral.isna().sum().sum()}")
print(f"População:    {df_populacao.isna().sum().sum()}")

print("\n=== REGISTROS DUPLICADOS ===")
print(f"Fraude 2025:  {df_fraude_2025.duplicated().sum()}")
print(f"Fraude Geral: {df_fraude_geral.duplicated().sum()}")
print(f"População:    {df_populacao.duplicated().sum()}")

print("\n=== AMOSTRA - FRAUDE 2025 ===")
display(df_fraude_2025.head(3))

print("\n=== AMOSTRA - POPULAÇÃO ===")
display(df_populacao.head(3))


=== DIMENSÕES DAS BASES ===
Fraude 2025:  52,461 registros | 12 colunas
Fraude Geral: 89,282 registros | 12 colunas
População:    234 registros | 4 colunas

=== PERÍODO DOS DADOS ===
Fraude Geral: 2023 a 2024
Fraude 2025: 2025 a 2025
População: 2023 a 2025

=== MUNICÍPIOS ===
Fraude 2025:  78
Fraude Geral: 78
População:    78

=== VALORES AUSENTES ===
Fraude 2025:  0
Fraude Geral: 0
População:    0

=== REGISTROS DUPLICADOS ===
Fraude 2025:  751
Fraude Geral: 676
População:    0

=== AMOSTRA - FRAUDE 2025 ===


,data_fato,hora_fato,data_criacao,tipo_boletim,tipo_incidente,municipio,bairro,tipo_local,ano,mes,mes_ano,ambiente_web
0,01/01/2025,14:17:00,01/01/2025,BU,CRIMES CONTRA PATRIMÔNIO: ESTELIONATO/FRAUDE,CARIACICA,VILA INDEPENDENCIA,VIA PÚBLICA,2025,1,01/2025,0
1,01/01/2025,22:54:00,02/01/2025,BABU,CRIMES CONTRA PATRIMÔNIO: ESTELIONATO/FRAUDE,SERRA,SERRA CENTRO,VIA PÚBLICA,2025,1,01/2025,0
2,01/01/2025,10:40:00,02/01/2025,BU,CRIMES CONTRA PATRIMÔNIO: ESTELIONATO/FRAUDE,VITORIA,MARIA ORTIZ,AMBIENTE WEB,2025,1,01/2025,1



=== AMOSTRA - POPULAÇÃO ===


,codigo_ibge,municipio,ano,populacao
0,102,Afonso Cláudio,2023,30684
1,102,Afonso Cláudio,2024,32446
2,102,Afonso Cláudio,2025,32456


# Passo 1: Unificação e Limpeza Fina.

In [ ]:
print("--- Passo 1: Unificação e Limpeza Fina ---")

import unicodedata

# 1. Função para padronizar nomes de municípios
def normalizar_municipio(valor):
    valor = str(valor).strip().upper()

    valor = unicodedata.normalize(
        'NFKD',
        valor
    )

    valor = ''.join(
        caractere
        for caractere in valor
        if not unicodedata.combining(caractere)
    )

    return valor


# 2. Unir as bases de fraude (2023/2024 + 2025)
df_fraude_completo = pd.concat(
    [
        df_fraude_geral,
        df_fraude_2025
    ],
    ignore_index=True
)


# 3. Corrigir o nome da coluna 'mes ' para 'mes'
df_fraude_completo.rename(
    columns={
        'mes ': 'mes'
    },
    inplace=True
)


# 4. Identificar registros com valores repetidos
duplicados_identificados = (
    df_fraude_completo
    .duplicated()
    .sum()
)

# Os registros repetidos são mantidos porque a base
# não possui identificador único da ocorrência.


# 5. Padronizar os nomes dos municípios
df_fraude_completo['municipio'] = (
    df_fraude_completo['municipio']
    .apply(normalizar_municipio)
)

df_populacao['municipio'] = (
    df_populacao['municipio']
    .apply(normalizar_municipio)
)


# 6. Validar compatibilidade dos municípios
municipios_fraude = set(
    df_fraude_completo['municipio']
    .unique()
)

municipios_populacao = set(
    df_populacao['municipio']
    .unique()
)

municipios_sem_populacao = (
    municipios_fraude
    - municipios_populacao
)

municipios_sem_ocorrencias = (
    municipios_populacao
    - municipios_fraude
)


# 7. Validar os resultados
print(
    f"Dimensão original (soma das bases): "
    f"{len(df_fraude_geral) + len(df_fraude_2025)} linhas"
)

print(
    f"Registros com valores repetidos identificados: "
    f"{duplicados_identificados}"
)

print(
    "Registros repetidos removidos: 0 "
    "(mantidos por ausência de identificador único)"
)

print(
    f"Dimensão final da base unificada: "
    f"{df_fraude_completo.shape}"
)

print(
    f"Anos presentes: "
    f"{sorted(df_fraude_completo['ano'].unique().tolist())}"
)

print(
    f"Municípios na base de ocorrências: "
    f"{len(municipios_fraude)}"
)

print(
    f"Municípios na base populacional: "
    f"{len(municipios_populacao)}"
)

print(
    "Municípios presentes nas ocorrências "
    "e ausentes na população:",
    municipios_sem_populacao
)

print(
    "Municípios presentes na população "
    "e ausentes nas ocorrências:",
    municipios_sem_ocorrencias
)

print(
    f"Colunas atuais: "
    f"{df_fraude_completo.columns.tolist()}"
)

print("\nAmostra dos dados (Head):")
display(
    df_fraude_completo.head(3)
)


Nesta etapa, as bases de ocorrências referentes aos anos de 2023, 2024 e 2025 foram unificadas em uma única estrutura.

Durante a preparação dos dados, o nome da coluna do mês foi corrigido de `'mes '` para `'mes'`, evitando inconsistências nas etapas seguintes. Também foram identificados registros com valores repetidos. Como a base disponibilizada não possui um identificador único de ocorrência que permita distinguir com segurança duplicatas reais de ocorrências distintas com atributos coincidentes, esses registros foram preservados.

Além disso, os nomes dos municípios foram normalizados por meio da remoção de espaços nas extremidades, conversão para letras maiúsculas e remoção de acentuação. Essa padronização cria uma chave textual consistente para integração entre a base de ocorrências e a base populacional.


# Passo 2: Construção da Grade Espaço-Temporal e Agrupamento

In [ ]:
print("--- Passo 2: Construção da Grade Espaço-Temporal ---")

# 1. Agrupar o total de ocorrências por município, ano e mês
df_ocorrencias_mensais = (
    df_fraude_completo
    .groupby(
        ['ano', 'mes', 'municipio'],
        as_index=False
    )
    .agg(
        total_ocorrencias=('municipio', 'size'),
        ocorrencias_web=('ambiente_web', 'sum')
    )
)

# 2. Criar a lista completa de municípios
municipios = sorted(
    df_populacao['municipio']
    .dropna()
    .unique()
)

# 3. Criar todas as combinações entre:
#    78 municípios x 3 anos x 12 meses
grade_completa = pd.MultiIndex.from_product(
    [
        [2023, 2024, 2025],
        range(1, 13),
        municipios
    ],
    names=[
        'ano',
        'mes',
        'municipio'
    ]
).to_frame(index=False)

# 4. Integrar as ocorrências reais à grade completa
df_agrupado = grade_completa.merge(
    df_ocorrencias_mensais,
    on=[
        'ano',
        'mes',
        'municipio'
    ],
    how='left'
)

# 5. Combinações sem registros de ocorrência
#    representam meses com zero ocorrências
df_agrupado[
    [
        'total_ocorrencias',
        'ocorrencias_web'
    ]
] = (
    df_agrupado[
        [
            'total_ocorrencias',
            'ocorrencias_web'
        ]
    ]
    .fillna(0)
    .astype(int)
)

# 6. Ordenar cronologicamente por município, ano e mês
df_agrupado = (
    df_agrupado
    .sort_values(
        by=['municipio', 'ano', 'mes']
    )
    .reset_index(drop=True)
)

# 7. Validar os resultados
total_grade = (
    df_agrupado['total_ocorrencias']
    .sum()
)

total_base = len(df_fraude_completo)

print(
    f"Total de combinações na grade: "
    f"{len(df_agrupado)} linhas (Municípios x Meses)"
)

print(
    f"Existem valores nulos na contagem? "
    f"{df_agrupado['total_ocorrencias'].isnull().any()}"
)

print(
    f"Total de ocorrências na base unificada: "
    f"{total_base}"
)

print(
    f"Total de ocorrências na grade: "
    f"{total_grade}"
)

print(
    f"Total de crimes agrupados bate com a base original? "
    f"{total_grade == total_base}"
)

print("\nAmostra da Grade Espaço-Temporal (Head):")
display(df_agrupado.head())

print("\nVerificando meses com 'Zero' ocorrências (Exemplo):")
display(
    df_agrupado[
        df_agrupado['total_ocorrencias'] == 0
    ].head(3)
)


Nesta etapa, foi construída a grade espaço-temporal contendo todas as combinações entre os 78 municípios e os 36 meses do período analisado, totalizando 2.808 linhas.

As ocorrências foram previamente agrupadas por município, ano e mês e, em seguida, integradas à grade completa. Combinações sem registros foram preenchidas com zero para preservar a continuidade das séries temporais.

A validação compara a soma de `total_ocorrencias` na grade com a quantidade de registros da base unificada. O resultado esperado é `True`, indicando que nenhuma ocorrência foi perdida durante o agrupamento e a construção da grade.


# Passo 3: Integração Populacional e Criação da Variável-Alvo.

In [ ]:
print("--- Passo 3: Integração Populacional e Criação da Variável-Alvo ---")

# 1. Realizar o Merge (Left Join) da grade com a base de população
# Usamos 'municipio' e 'ano' como chaves para garantir que a população mude a cada ano
df_final = df_agrupado.merge(df_populacao[['municipio', 'ano', 'populacao']],
                             on=['municipio', 'ano'],
                             how='left')

# 2. Calcular a Taxa Mensal por 100 mil habitantes (Variável-alvo principal)
# Fórmula da metodologia: Taxa100k = (Ocorrências mensais / População anual) * 100.000
df_final['taxa_100k_total'] = (df_final['total_ocorrencias'] / df_final['populacao']) * 100000

# 3. Calcular a Taxa Mensal apenas para crimes no Ambiente Web (Análise complementar)
df_final['taxa_100k_web'] = (df_final['ocorrencias_web'] / df_final['populacao']) * 100000

# 4. Arredondar as taxas para 2 casas decimais (melhora a visualização e modelagem)
df_final['taxa_100k_total'] = df_final['taxa_100k_total'].round(2)
df_final['taxa_100k_web'] = df_final['taxa_100k_web'].round(2)

# 5. Validar os resultados
print(f"Dimensão após integração populacional: {df_final.shape}")
print(f"Existem municípios sem população mapeada (NaN)? {df_final['populacao'].isnull().sum()} registros vazios")

print("\nAmostra dos Dados com a Taxa Calculada (Head):")
display(df_final.head())

print("\nEstatísticas Descritivas das Taxas:")
display(df_final[['taxa_100k_total', 'taxa_100k_web']].describe())

Nesta etapa, a grade espaço-temporal é integrada às estimativas anuais de população utilizando município e ano como chaves. A validação de valores ausentes permite verificar se todos os municípios e anos foram corretamente associados à população correspondente.

A partir dessa integração são calculadas as taxas mensais de ocorrências por 100 mil habitantes, tanto para o conjunto total de registros quanto para aqueles classificados como relacionados ao ambiente Web. Essas taxas constituem as principais variáveis utilizadas nas análises e na modelagem posterior.


# Passo 4: Engenharia de Características

In [ ]:
print("--- Passo 4: Engenharia de Características (Lags e Médias Móveis) ---")

# 1. Garantir que os dados estão perfeitamente ordenados no tempo
df_final = (
    df_final
    .sort_values(
        by=['municipio', 'ano', 'mes']
    )
    .reset_index(drop=True)
)

# 2. Criar as variáveis de Lag (Taxas dos meses anteriores)
# O shift(1) "empurra" o dado do mês anterior para a linha do mês atual
df_final['taxa_lag_1'] = (
    df_final
    .groupby('municipio')['taxa_100k_total']
    .shift(1)
)

df_final['taxa_lag_2'] = (
    df_final
    .groupby('municipio')['taxa_100k_total']
    .shift(2)
)

df_final['taxa_lag_3'] = (
    df_final
    .groupby('municipio')['taxa_100k_total']
    .shift(3)
)

# 3. Criar as Médias Móveis dos últimos 3 e 6 meses
# As médias utilizam apenas meses anteriores ao período atual.
# min_periods=3 e min_periods=6 garantem que a média só seja
# calculada quando existir o histórico completo da respectiva janela.
df_final['media_movel_3m'] = (
    df_final
    .groupby('municipio')['taxa_lag_1']
    .transform(
        lambda x: x.rolling(
            window=3,
            min_periods=3
        ).mean()
    )
)

df_final['media_movel_6m'] = (
    df_final
    .groupby('municipio')['taxa_lag_1']
    .transform(
        lambda x: x.rolling(
            window=6,
            min_periods=6
        ).mean()
    )
)

# 4. Arredondar os novos campos para 2 casas decimais
colunas_novas = [
    'taxa_lag_1',
    'taxa_lag_2',
    'taxa_lag_3',
    'media_movel_3m',
    'media_movel_6m'
]

df_final[colunas_novas] = (
    df_final[colunas_novas]
    .round(2)
)

# 5. Validar os resultados
print(
    f"Colunas atuais na base: "
    f"{df_final.columns.tolist()}"
)

print(
    "Total de valores nulos gerados "
    "(esperado devido aos primeiros meses sem histórico):"
)

print(
    df_final[colunas_novas]
    .isnull()
    .sum()
)

print(
    "\nAmostra dos Dados com as Variáveis Temporais "
    "(Focando em um município para ver a evolução):"
)

display(
    df_final[
        [
            'municipio',
            'ano',
            'mes',
            'taxa_100k_total',
            'taxa_lag_1',
            'taxa_lag_2',
            'taxa_lag_3',
            'media_movel_3m',
            'media_movel_6m'
        ]
    ].head(8)
)


Nesta etapa, foram construídas variáveis temporais utilizando exclusivamente informações anteriores ao mês que se deseja estimar.

Os lags de 1, 2 e 3 meses representam diretamente as taxas observadas nos períodos anteriores. Já as médias móveis de 3 e 6 meses são calculadas sobre a taxa defasada em um mês, impedindo que o valor do próprio período-alvo seja utilizado na construção da feature.

Para manter o significado temporal das variáveis, a média móvel de 3 meses só é calculada quando existem três meses anteriores disponíveis, enquanto a média móvel de 6 meses exige seis meses completos de histórico. Por esse motivo, são esperados valores nulos nos primeiros meses de cada município.

Com 78 municípios, os valores nulos esperados são:
- `taxa_lag_1`: 78;
- `taxa_lag_2`: 156;
- `taxa_lag_3`: 234;
- `media_movel_3m`: 234;
- `media_movel_6m`: 468.

Esses valores são consequência natural da construção das séries temporais e não representam dados faltantes na base original. As linhas sem histórico suficiente serão excluídas apenas na etapa de modelagem.


# Passo 5: Análise Exploratória dos Dados

In [ ]:
print("--- Passo 5: Análise Exploratória dos Dados (Visualização) ---")
import matplotlib.pyplot as plt
import seaborn as sns

# Configuração visual padrão para os gráficos
sns.set_theme(style="whitegrid")

# 1. Preparar dados para o gráfico de evolução estadual
# Agrupamos por ano e mês tirando a média das taxas de todos os municípios
df_evolucao = df_final.groupby(['ano', 'mes'])[['taxa_100k_total', 'taxa_100k_web']].mean().reset_index()

# Criar uma coluna de 'Período' no formato AAAA-MM para facilitar a leitura no eixo X
df_evolucao['periodo'] = df_evolucao['ano'].astype(str) + '-' + df_evolucao['mes'].astype(str).str.zfill(2)

# --- FIGURA 1: Evolução Temporal ---
plt.figure(figsize=(14, 6))
plt.plot(df_evolucao['periodo'], df_evolucao['taxa_100k_total'], marker='o', label='Taxa Total (Média ES)', linewidth=2)
plt.plot(df_evolucao['periodo'], df_evolucao['taxa_100k_web'], marker='s', label='Taxa Web (Média ES)', linewidth=2, linestyle='--')
plt.xticks(rotation=45)
plt.title('Evolução Mensal da Taxa de Estelionato e Fraude por 100 mil hab. (Média dos Municípios)')
plt.xlabel('Período (Ano-Mês)')
plt.ylabel('Taxa Média por 100 mil hab.')
plt.legend()
plt.tight_layout()
plt.show()

# --- FIGURA 2: Top 10 Municípios (Ambiente Web) ---
# Calcular a média histórica da taxa web por município
top10_web = df_final.groupby('municipio')['taxa_100k_web'].mean().sort_values(ascending=False).head(10).reset_index()

plt.figure(figsize=(12, 6))
sns.barplot(
    data=top10_web,
    x='taxa_100k_web',
    y='municipio',
    hue='municipio',
    palette='viridis',
    legend=False
)
plt.title('Top 10 Municípios - Maior Taxa Média Mensal de Fraudes na Web (2023-2025)')
plt.xlabel('Taxa Média por 100 mil hab.')
plt.ylabel('Município')
plt.tight_layout()
plt.show()

# --- FIGURA 3: Boxplot para Identificar Outliers ---
plt.figure(figsize=(10, 5))
sns.boxplot(x=df_final['taxa_100k_total'], color='lightcoral')
plt.title('Distribuição da Taxa Mensal (Total) e Identificação de Valores Discrepantes (Outliers)')
plt.xlabel('Taxa Mensal por 100 mil hab.')
plt.tight_layout()
plt.show()

print("Gráficos gerados com sucesso! Analise as imagens para insights metodológicos.")


Nesta etapa, são geradas visualizações exploratórias para analisar a evolução temporal das ocorrências, a distribuição das taxas normalizadas e as diferenças entre municípios.

Os gráficos têm como objetivo apoiar a identificação de tendências, sazonalidade, valores discrepantes e localidades com taxas mais elevadas, servindo de base para interpretação dos dados e para as decisões de modelagem.

A visualização dos municípios com maiores taxas relacionadas ao ambiente Web utiliza o próprio município como variável de agrupamento (`hue`) apenas para compatibilidade com a API atual do Seaborn, sem alterar o conteúdo analítico do gráfico.


# Passo 6: Modelagem

In [ ]:
print("--- Passo 6: Modelagem (Treino, Teste e Avaliação) ---")
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np
import pandas as pd

# 1. Tratar valores nulos gerados pelos Lags (não podemos treinar com NaN)
df_model = df_final.dropna(subset=['taxa_lag_3', 'media_movel_6m']).copy()

# 2. Transformar a variável 'municipio' em colunas numéricas (One-Hot Encoding)
# Isso permite que os modelos interpretem a dimensão espacial (qual cidade é qual)
df_model = pd.get_dummies(df_model, columns=['municipio'], drop_first=True)

# 3. Divisão Temporal Rígida (Sem vazamento de dados)
df_train = df_model[df_model['ano'] < 2025].copy() # 2023 e 2024
df_test = df_model[df_model['ano'] == 2025].copy()  # 2025

# Definir as colunas de entrada (Features - X) e a variável-alvo (Target - y)
colunas_excluir = ['ano', 'total_ocorrencias', 'ocorrencias_web', 'taxa_100k_total', 'taxa_100k_web']
features = [col for col in df_model.columns if col not in colunas_excluir]

X_train = df_train[features]
y_train = df_train['taxa_100k_total']
X_test = df_test[features]
y_test = df_test['taxa_100k_total']

# 4. Construir o Baseline: "Amanhã será igual a hoje" (taxa do mês anterior)
y_pred_baseline = df_test['taxa_lag_1']

# 5. Treinar os Algoritmos de Machine Learning
# Random Forest
print("Treinando Random Forest...")
rf_model = RandomForestRegressor(random_state=42, n_estimators=100)
rf_model.fit(X_train, y_train)
y_pred_rf = rf_model.predict(X_test)

# XGBoost
print("Treinando XGBoost...")
xgb_model = XGBRegressor(random_state=42, n_estimators=100)
xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

# 6. Avaliar os Resultados
def avaliar_modelo(nome, y_true, y_pred):
    mae = mean_absolute_error(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    r2 = r2_score(y_true, y_pred)
    print(f"{nome:<25} -> MAE: {mae:.2f} | RMSE: {rmse:.2f} | R2: {r2:.4f}")

print("\n--- Desempenho no Período de Teste (2025) ---")
avaliar_modelo("Baseline (Mês Anterior)", y_test, y_pred_baseline)
avaliar_modelo("Random Forest", y_test, y_pred_rf)
avaliar_modelo("XGBoost", y_test, y_pred_xgb)

Analisando as métricas obtidas no teste com os dados de 2025:

Ganho Preditivo Comprovado: O Baseline (que assume que o mês atual será igual ao mês anterior) obteve o maior erro médio absoluto (MAE: 22.71) e explicou apenas 28% da variância das taxas ($R^2$: 0.2826). Isso comprova a premissa da sua metodologia de que métodos baseados exclusivamente no histórico imediato são insuficientes.  

O Vencedor (Random Forest): O Random Forest teve o melhor desempenho geral, reduzindo o erro médio (MAE) para 19.97 e o RMSE para 26.42. O salto no coeficiente de determinação ($R^2$) para 0.4622 é o grande destaque aqui. Na literatura de predição criminal, que lida com fenômenos sociais de alta complexidade e ruído, conseguir explicar mais de 46% da variância espacial e temporal apenas com dados históricos é um resultado estatisticamente muito sólido.  


Desempenho do XGBoost: O modelo obteve resultados intermediários (MAE: 21.32, $R^2$: 0.3932). Embora robusto, ele costuma ser mais sensível a hiperparâmetros não otimizados. Para o escopo deste artigo, o Random Forest em sua configuração padrão já entregou a capacidade preditiva necessária.

# Passo 7: Demonstração e Comunicação

## Passo 8: Exportação dos Artefatos para o MVP

Nesta etapa, são salvos os principais artefatos necessários para a integração com o MVP em Streamlit. O objetivo é evitar que a aplicação precise refazer todo o treinamento a cada execução.

São exportados:
- a base processada final, com taxas e variáveis temporais;
- o modelo Random Forest selecionado como melhor modelo entre os avaliados;
- a lista de colunas utilizadas durante o treinamento, garantindo consistência na etapa de inferência.


In [ ]:
print("--- Passo 8: Exportação dos Artefatos para o MVP ---")

import json
import joblib
from pathlib import Path

# 1. Definir as pastas de saída dentro do repositório
PROJECT_ROOT = Path('/content/ml-cybercrime-public-safety')

DATA_PROCESSED_OUTPUT = (
    PROJECT_ROOT
    / 'data'
    / 'processed'
    / 'modelagem'
)

MODELS_OUTPUT = (
    PROJECT_ROOT
    / 'models'
)

DATA_PROCESSED_OUTPUT.mkdir(
    parents=True,
    exist_ok=True
)

MODELS_OUTPUT.mkdir(
    parents=True,
    exist_ok=True
)

# 2. Salvar a base processada final
caminho_base_final = (
    DATA_PROCESSED_OUTPUT
    / 'painel_municipal_mensal.csv'
)

df_final.to_csv(
    caminho_base_final,
    index=False
)

# 3. Salvar o modelo Random Forest selecionado
caminho_modelo_rf = (
    MODELS_OUTPUT
    / 'random_forest.joblib'
)

joblib.dump(
    rf_model,
    caminho_modelo_rf
)

# 4. Salvar a lista de features utilizadas pelo modelo
caminho_features = (
    MODELS_OUTPUT
    / 'features_modelo.json'
)

with open(
    caminho_features,
    'w',
    encoding='utf-8'
) as arquivo:
    json.dump(
        features,
        arquivo,
        ensure_ascii=False,
        indent=2
    )

# 5. Validar os arquivos gerados
print(
    f"Base processada salva em: "
    f"{caminho_base_final}"
)

print(
    f"Modelo Random Forest salvo em: "
    f"{caminho_modelo_rf}"
)

print(
    f"Lista de features salva em: "
    f"{caminho_features}"
)

print(
    f"Total de linhas na base exportada: "
    f"{len(df_final)}"
)

print(
    f"Total de features do modelo: "
    f"{len(X_train.columns)}"
)

print("\nArtefatos exportados com sucesso!")


In [ ]:
print("--- Passo 7: Demonstração e Comunicação (Resultados 2025) ---")

# 1. Recuperar as informações originais (município e mês) preservando o índice do teste
resultados = df_final.loc[df_test.index, ['municipio', 'mes', 'taxa_100k_total']].copy()
resultados.rename(columns={'taxa_100k_total': 'taxa_observada'}, inplace=True)

# 2. Inserir as previsões do Random Forest (nosso melhor modelo)
resultados['taxa_prevista'] = y_pred_rf.round(2)

# 3. Calcular a Diferença Absoluta e a Variação (conforme metodologia)
# Diferença absoluta mede o tamanho do erro.
resultados['diferenca_absoluta'] = abs(resultados['taxa_observada'] - resultados['taxa_prevista']).round(2)

# Variação mede o sentido do desvio:
# Positivo = Município teve MAIS fraudes do que o modelo previu (Foco de atenção)
# Negativo = Município teve MENOS fraudes do que o modelo previu
resultados['variacao'] = (resultados['taxa_observada'] - resultados['taxa_prevista']).round(2)

# 4. Mostrar os casos mais atípicos (Maiores desvios absolutos)
print("\n--- Comportamentos Atípicos em 2025 (Top 5 Diferenças Absolutas) ---")
atipicos = resultados.sort_values(by='diferenca_absoluta', ascending=False).head(5)
display(atipicos)

# 5. Criar o Ranking de Municípios por Risco Previsto (Média de 2025)
print("\n--- Ranking de Taxa Prevista (Top 10 Municípios em 2025) ---")
ranking_risco = resultados.groupby('municipio')['taxa_prevista'].mean().sort_values(ascending=False).head(10).reset_index()
display(ranking_risco)

print("\nConcluído! Estes dados servem de base para a construção do painel visual citado no artigo.")


1. A Força dos Comportamentos Atípicos:
Os maiores desvios absolutos aconteceram em cidades menores, como Águia Branca, Iconha e Mantenópolis. Isso faz todo sentido estatístico: em municípios com população reduzida, um único ataque cibernético ou fraude em massa (como o vazamento de dados de clientes de uma loja local) pode gerar uma onda súbita de dezenas de boletins de ocorrência em um único mês. Como a população é pequena, a taxa por 100 mil habitantes explode. O Random Forest, baseando-se no histórico calmo dessas cidades, previu taxas baixas (na casa dos 40 a 60), evidenciando o quão anômalos foram os picos observados (acima de 170). Isso cumpre rigorosamente a premissa da sua avaliação (Seção 4.5): o modelo serve para sinalizar esses desvios como alertas, e não necessariamente para indicar uma causa criminal específica logo de cara.  

2. A Consistência do Ranking de Risco: O Top 10 de risco previsto em 2025 mostra que o modelo aprendeu brilhantemente a geografia e a demografia criminal do Espírito Santo. Vitória (156.16) e Vila Velha (146.58) lideram o ranking, seguidas por grandes polos turísticos litorâneos (Guarapari e Marataízes) e centros econômicos regionais (Linhares, Serra e Cachoeiro de Itapemirim). A Inteligência Artificial conseguiu compreender que a dinâmica de fraudes é estruturalmente mais intensa nessas regiões.

# Conclusão

### Conclusão do Estudo de Predição Espaço-Temporal

Esta pesquisa demonstrou com sucesso a viabilidade de utilizar técnicas de *Machine Learning* para estimar as taxas mensais de estelionato e fraude nos municípios do Espírito Santo. Através da integração de dados criminais públicos com estimativas populacionais do IBGE, foi possível normalizar as ocorrências e realizar uma comparação justa entre cidades de diferentes portes.

**Principais Resultados Obtidos:**

1. **Superioridade da Inteligência Artificial:** O modelo **Random Forest Regressor** apresentou o melhor desempenho preditivo, superando tanto o modelo *XGBoost* quanto o método *Baseline* (que utilizava apenas a taxa do mês anterior). A redução do Erro Médio Absoluto (MAE) e o aumento significativo do coeficiente de determinação ($R^2$) comprovam que o histórico criminal de 1 a 6 meses carrega padrões que a IA consegue identificar para prever o futuro.
2. **Mapeamento de Risco:** O algoritmo foi capaz de compreender a dinâmica estrutural da criminalidade, projetando o maior risco continuado para os grandes centros urbanos (Vitória, Vila Velha, Serra) e polos turísticos (Guarapari, Marataízes).
3. **Detecção de Comportamentos Atípicos:** A variação absoluta entre o valor previsto pelo modelo e o valor real observado mostrou-se uma ferramenta valiosa para identificar anomalias (como explosões de ocorrências em Águia Branca, Iconha e Mantenópolis). Esses picos evidenciam a mudança do paradigma criminal, onde cidades pequenas, antes consideradas seguras contra crimes patrimoniais clássicos, tornam-se repentinamente alvos de ataques cibernéticos em massa.

**Impacto na Segurança Pública:**
O *pipeline* desenvolvido consolida a transição de um modelo de segurança pública reativo para uma abordagem de **inteligência preventiva**. Com estimativas confiáveis do risco de fraudes por município, o Estado pode alocar recursos de investigação de crimes cibernéticos e direcionar campanhas de educação digital de forma estratégica, atuando antes que a tendência de alta se torne um dano irreversível para a população.

## Resultados consolidados da modelagem

Com o pipeline corrigido e reexecutado integralmente, os resultados oficiais do conjunto de teste de 2025 são:

| Modelo | MAE | RMSE | R² |
|---|---:|---:|---:|
| Baseline (mês anterior) | 23.28 | 32.21 | 0.2735 |
| Random Forest | **20.21** | **27.96** | **0.4527** |
| XGBoost | 21.54 | 29.45 | 0.3930 |

O Random Forest apresentou o melhor desempenho geral entre as abordagens avaliadas. Os maiores desvios entre valores observados e previstos devem ser interpretados como comportamentos atípicos em relação ao padrão aprendido pelo modelo, e não como evidência automática de uma causa criminal específica.
